source/extensions/access_loggers/common/grpc_access_logger.h

```diff
@@ -146,6 +160,84 @@ template <typename LogRequest, typename LogResponse> class GrpcAccessLogClient {
   const absl::optional<envoy::config::core::v3::RetryPolicy> grpc_stream_retry_policy_;
 };
 
+constexpr absl::string_view GRPC_LOG_STATS_PREFIX = "access_logs.grpc_access_log.";
+
+#define CRITICAL_ACCESS_LOGGER_GRPC_CLIENT_STATS(COUNTER, GAUGE)                                   \
+  COUNTER(critical_logs_nack_received)                                                             \
+  COUNTER(critical_logs_ack_received)
+
+struct GrpcCriticalAccessLogClientGrpcClientStats {
+  CRITICAL_ACCESS_LOGGER_GRPC_CLIENT_STATS(GENERATE_COUNTER_STRUCT, GENERATE_GAUGE_STRUCT)
+};
+
+template <class RequestType, class ResponseType> class GrpcCriticalAccessLogClient {
+public:
+  struct CriticalLogStreamCallbacks : public Grpc::AsyncStreamCallbacks<ResponseType> {
+    explicit CriticalLogStreamCallbacks(GrpcCriticalAccessLogClient& parent) : parent_(parent) {}
+
+    // Grpc::AsyncStreamCallbacks
+    void onCreateInitialMetadata(Http::RequestHeaderMap&) override {}
+    void onReceiveInitialMetadata(Http::ResponseHeaderMapPtr&&) override {}
+    void onReceiveMessage(std::unique_ptr<ResponseType>&& message) override {
+      const auto& id = message->id();
+
+      switch (message->status()) {
+      case envoy::service::accesslog::v3::CriticalAccessLogsResponse::ACK:
+        parent_.stats_.critical_logs_ack_received_.inc();
+        parent_.buffered_client_.onSuccess(id);
+        break;
+      case envoy::service::accesslog::v3::CriticalAccessLogsResponse::NACK:
+        parent_.stats_.critical_logs_nack_received_.inc();
+        parent_.buffered_client_.onError(id);
+        break;
+      default:
```

Comment:
You should remove this default, such that compiler generated an error when all cases in enum were not covered.